src/Nethermind/Nethermind.Evm/TransactionSubstate.cs

```diff
@@ -41,30 +42,35 @@ public TransactionSubstate(EvmExceptionType exceptionType, bool isTracerConnecte
             ShouldRevert = shouldRevert;
             if (ShouldRevert)
             {
-                // TODO: is this invoked even if there is no tracer? why would we construct error messages then?
                 Error = Revert;
                 if (isTracerConnected)
                 {
                     if (Output.Length > 0)
                     {
-                        try
-                        {
-                            BigInteger start = Output.Span.Slice(4, 32).ToUnsignedBigInteger();
-                            BigInteger length = Output.Slice((int)start + 4, 32).Span.ToUnsignedBigInteger();
-                            Error = string.Concat("Reverted ",
-                                Output.Slice((int)start + 32 + 4, (int)length).ToArray().ToHexString(true));
-                        }
-                        catch (Exception)
+                        ReadOnlySpan<byte> span = Output.Span;
+                        if (span.Length >= 32 * 2 + 4)
                         {
                             try
                             {
-                                Error = string.Concat("Reverted ", Output.ToArray().ToHexString(true));
+                                int start = (int)(new UInt256(span.Slice(4, 32), isBigEndian: true));
+                                if (start + 4 + 32 <= span.Length)
+                                {
+                                    int length = (int)new UInt256(span.Slice(start + 4, 32), isBigEndian: true);
+                                    if (checked(start + 4 + 32 + length) <= span.Length)
+                                    {
+                                        Error = string.Concat("Reverted ",
+                                            span.Slice(start + 32 + 4, length).ToHexString(true));
+                                        return;
+                                    }
+                                }
```

Comment:
would love to see const used for those numbers `4` and `32`'s

Reply:
Changed